### PR TITLE
passthrough examples: enable direct io when open has flag: O_DIRECT

### PR DESCRIPTION
**Pull request by yangyun50**
I use the passthrough  example (i.e., passthrough_hp) as the fuse daemon. In the application, when I open a file with flag O_DIRECT and read some data from the file, it doesn't go through into the `fuse_direct_read_iter `in the fuse kernel. The reason is that the fuse daemon dosen't check the O_DIRECT flag and enable it. 
So enable it when flag O_DIRECT is set.

### CODE REVIEW

**Comment by trapexit** [comment]
direct io and O_DIRECT are different things and AFAIK should be kept separated in the library. It is the servers responsibly to manage that as it sees fit.

**Comment by bsbernd** [comment]
The patch itself is indeed likely going to happen, just not for the reason in the commit message. @trapexit pointed out FOPEN_DIRECT_IO and O_DIRECT are not entirely the same. However, in order to get a shared lock  for writes to the same file FOPEN_DIRECT_IO is needed. I have a patch series that consolidates kernel side write path for FOPEN_DIRECT_IO and O_DIRECT, but as it turnes out, we actually still cannot enable the shared lock for plain O_DIRECT - reason is that one can set that with fcntl, but for the shared lock we have to be absolutely sure there are no dirty pages going to be written out. 
So the patch itself is ok, but deserves the right commit message and also a comment in the code explaining the reason.   Also, the reason in the commit message - going through fuse_direct_read_iter() is no good reason - fuse_cache_read_iter() -> generic_file_read_iter() perfectly handles O_DIRECT.

Link that explains why FOPEN_DIRECT_IO will be needed (this is an ongoing patch set)
https://lore.kernel.org/all/753d6823-e984-4730-a126-d66b65ea772c@ddn.com/

Patch series to consolidate direct IO
https://lwn.net/ml/linux-fsdevel/20230918150313.3845114-1-bschubert@ddn.com/

v5 depends on the patch series in the discussion above - will be posted as dependency.

So in summary, please do not ask for the patch to merged with the current commit - needs the shared lock explanation, reference to the link above and also a short comment in the code.

**Comment by yangyun50** [comment]
Thanks  for your advice. I am so sorry for making a bad pull request since I am not so familiar with the differences between FOPEN_DIRECT_IO and O_DIRECT.  And it needs some time for me to figure it out.  So I just delay this work until I have time.

**Comment by bsbernd** [comment]
Don't worry, I will create a pull with the right explanation and add you as reviewer. I'm just waiting for kernel side patches to be merged (they are in fuse-next for now).

**Comment by yangyun50** [comment]
ok, thanks!

**Comment by yangyun50** [comment]
> Don't worry, I will create a pull with the right explanation and add you as reviewer. I'm just waiting for kernel side patches to be merged (they are in fuse-next for now).

Hi, bsbernd. Which kernel side patches do you mean, the [consolidate direct IO](https://lwn.net/ml/linux-fsdevel/20230918150313.3845114-1-bschubert@ddn.com/)  patches or   [fuse: inode IO modes and mmap](https://lore.kernel.org/all/20240131230827.207552-1-bschubert@ddn.com/#r) patches ? Could you show me the kernel sid patches link in fuse-next ? I haven't found it. Thanks a lot.

**Comment by yangyun50** [comment]
I have read the patches you provided and the kernel code about FOPEN_DIRECT_IO.  Based on this,  the new commits are created  with detailed explanation, please review it. Thank you.

**Comment by yangyun50** [comment]
The test seems to be failed for a reason unrelated to this PR.

**Review by bsbernd — COMMENTED** [review]
@yangyun50  Thank you for the update!

I think this part of the comment is not needed anymore with latest IO mode patches.

```
Patch series to consolidate the FOPEN_DIRECT_IO and O_DIRECT
https://lwn.net/ml/linux-fsdevel/20230918150313.3845114-1-bschubert@ddn.com
```

And regarding IO mode patches - the actual reason why we cannot enable the shared lock for O_DIRECT is that one can set that with fcntl(F_SETFD, ...) - all the precautions we are doing in the IO-mode patches are worked around with that. This is the important part that is missing in the commit message. Can you please add something like

```
Shared locks (FOPEN_PARALLEL_DIRECT_WRITES) cannot be enabled for O_DIRECT, as O_DIRECT may be set
past file open time with fcntl(fd, F_SETFD, ...). Kernel side fuse has precautions for shared lock
direct-IO, as it needs an exclusive inode lock when direct and page cache IO happen at the
the time. 
``

**Comment by bsbernd** [comment]
@yangyun50 I know that some unit tests fail, I'm in the process to fix these.

File: example/passthrough_fh.c [hunk 1]
```diff
@@ -384,6 +384,12 @@ static int xmp_open(const char *path, struct fuse_file_info *fi)
 	if (fd == -1)
 		return -errno;
 
+        /* Enable direct_io when open has flags O_DIRECT to enjoy the feature
+           parallel_direct_writes (i.e., to get a shard lock (not exclusive lock)
+           for writes to the same file). */
+        if (fi->flags & O_DIRECT)
+		fi->direct_io = 1;
```
[the remark this answers is not in the document]
Comment (yangyun50) [reply example/passthrough_fh.c:391]: done

File: example/passthrough.c [hunk 2]
```diff
@@ -303,6 +303,12 @@ static int xmp_open(const char *path, struct fuse_file_info *fi)
 	if (res == -1)
 		return -errno;
 
+        /* Enable direct_io when open has flags O_DIRECT to enjoy the feature
+        parallel_direct_writes (i.e., to get a shard lock (not exclusive lock)
+        for writes to the same file). */
+	if (fi->flags & O_DIRECT)
+		fi->direct_io = 1;
+
```
[the remark this answers is not in the document]
Comment (yangyun50) [reply example/passthrough.c:313]: done

**Review by bsbernd — APPROVED** [review]

**Comment by bsbernd** [comment]
Thanks for your contribution, merged.

**Comment by bsbernd** [comment]
Maybe easier to ask by mail directly or fsdevel mail? Anyway, no I didn't have time for it. It also needs quite some rebasing after the passthrough changes. I think I first need to discuss with Miklos if we want to have these patches at all.
